src/Callable.h

```diff
@@ -0,0 +1,357 @@
+#ifndef HALIDE_CALLABLE_H
+#define HALIDE_CALLABLE_H
+
+/** \file
+ *
+ * Defines the front-end class representing a jitted, callable Halide pipeline.
+ */
+
+#include <array>
+#include <map>
+
+#include "Buffer.h"
+#include "IntrusivePtr.h"
+#include "JITModule.h"
+
+namespace Halide {
+
+struct Argument;
+struct CallableContents;
+
+namespace PythonBindings {
+class PyCallable;
+}
+
+namespace Internal {
+
+template<typename>
+struct IsHalideBuffer : std::false_type {};
+
+template<typename T, int Dims>
+struct IsHalideBuffer<::Halide::Buffer<T, Dims>> : std::true_type {};
+
+template<typename T, int Dims>
+struct IsHalideBuffer<::Halide::Runtime::Buffer<T, Dims>> : std::true_type {};
+
+template<>
+struct IsHalideBuffer<halide_buffer_t *> : std::true_type {};
+
+template<typename>
+struct HalideBufferStaticTypeAndDims {
+    static constexpr halide_type_t type() {
+        return halide_type_t();
+    }
+    static constexpr int dims() {
+        return -1;
+    }
+};
+
+template<typename T, int Dims>
+struct HalideBufferStaticTypeAndDims<::Halide::Buffer<T, Dims>> {
```

Comment:
Should these be static methods on Halide::Buffer and Halide::Runtime::Buffer instead?

Reply:
...maybe? I figured they were specialized enough not to warrant adding to both of those classes. Easy enough to generalize if we need them elsewhere someday (but so far, we haven't).